src/Effect/Logger.hs

```diff
@@ -78,21 +80,21 @@ logWarn = log SevWarn
 logError :: Has Logger sig m => Doc AnsiStyle -> m ()
 logError = log SevError
 
-withLogger :: MonadIO m => Severity -> LoggerC m a -> m a
+withLogger :: Has (Lift IO) sig m => Severity -> LoggerC m a -> m a
 withLogger minSeverity (LoggerC act) = do
-  isTerminal <- liftIO $ hIsTerminalDevice stderr
+  isTerminal <- sendIO $ hIsTerminalDevice stderr
   let logger = bool rawLogger termLogger isTerminal
-  queue <- liftIO (hSetBuffering stderr NoBuffering *> newTMQueueIO @LogMsg)
+  queue <- sendIO (hSetBuffering stderr NoBuffering *> newTMQueueIO @LogMsg)
 
-  tid <- liftIO $ async $ logger minSeverity queue
+  let mkLogger = sendIO $ async $ logger minSeverity queue
 
-  res <- runReader queue act
+      flushLogger tid = sendIO $ do
+        atomically $ closeTMQueue queue
+        void (wait tid)
 
-  liftIO $ do
-    atomically $ closeTMQueue queue
-    void (wait tid)
-
-  pure res
+  bracket mkLogger
```

Comment:
This now ensures that the log queue is flushed before exiting, even in the face of exceptions